proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+code snippets of how they will be called.
+
+### Out-of-process system initialization
+
+On first creation of a low-level compiler api instance, a singleton object is
+created in-process to the caller.
+
+The singleton object manages all state and request traffic from the calling
+process. The singleton contains an api dispatching system that interfaces with
+a process pool. Each worker process in the pool is monitored by a dedicated
+thread. One thread to one process.
+
+Worker monitoring threads use named pipes as IPC to communicate with their
+workers and a different method (TBD) to communicate with the singleton's
+api dispatching system.
+
+Worker processes exit unconditionally after completing work. The monitoring
+thread spawns a new process for any exited process to replenish the process
+pool. Exited processes include crashed processes.  Processes that crash are
+detected by the monitoring thread allowing error information to be communicated
+back through the system.
+
+The worker process (Clang.exe) will be updated to include additional command
+options that indicate it should be launched as a worker and establish the IPC
+mechanism with a monitoring thread owned by the Compiler Process Pool.
+
+## Architecture Diagram
+```mermaid
+flowchart TD
+    subgraph CompilerProcessPool["Compiler Process Pool"]
+      subgraph MonitorThread1["Monitor Thread (1)"]
+        subgraph JsonRPC ["IPC Dispatcher"]
+        end
+      end
+      subgraph MonitorThread2["Monitor Thread (2)"]
+        subgraph JsonRPC2 ["IPC Dispatcher"]
+        end
+      end
+    end
+
+    subgraph Process1["process 1"]
+        ClangProcess1["clang.exe (1)"]
+        db[("compiled<br/>shader")]
+    end
+
+    subgraph Process2["process 2"]
+        ClangProcess2["clang.exe (2)"]
+        db2[("compiled<br/>shader")]
+    end
+
+    subgraph HighLevelCApi["Compiler Api (High-Level)"]
+        HighLevelApiBehaviorLogic("implements sync/async policy, packs/unpacks parameters")
+    end
+
+    subgraph LowLevelCApi["Compiler Api (Low-Level)"]
+        LowLevelApi("communicates with process pool, manages inputs/results")
+    end
+
+    style Process1 stroke-dasharray:10,10
+    style Process2 stroke-dasharray:10,10
+  
+    ApiCallDispatcher{"Api Dispatcher<br/>(find available process and dispatch work)"}
+
+    Api["Compile/CompileAsync()"] --> HighLevelApiBehaviorLogic <--> LowLevelCApi <--> |"dispatch and monitor call"| ApiCallDispatcher
+
+    ApiCallDispatcher -->JsonRPC<--> | json-rpc<br/>protocol | ClangProcess1 --> db
+
+    ApiCallDispatcher --> JsonRPC2<-->| json-rpc<br/>protocol | ClangProcess2 --> db2
+```
+
+### Calling apis
+A high-level api call starts in the api entrypoint implementation which will
+call into low-level apis and implement sync/async behaviors. The low-level
+c-api is responsible for finding an available worker process (clang.exe) to
+perform the work.  If a worker is not available the high-level api behavior
```

Comment:
So I'm trying to understand the proposed design here since it is quite different to what I was expecting (see above). I can't quite understand what the difference is between the low- and high-level APIs in this design?

Reply:
Yes.  I changed this document.  It was going in multiple directions at the same time which was really confusing.  I removed lots of this type of content.